tensorstore/kvstore/s3/s3_resource.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright 2022 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tensorstore/kvstore/s3/s3_resource.h"
+
+#include <memory>
+#include <optional>
+
+#include "absl/base/call_once.h"
+#include "absl/flags/marshalling.h"
+#include "absl/log/absl_log.h"
+#include "absl/time/time.h"
+#include "tensorstore/context_resource_provider.h"
+#include "tensorstore/internal/env.h"
+#include "tensorstore/kvstore/gcs_http/admission_queue.h"
+#include "tensorstore/kvstore/gcs_http/scaling_rate_limiter.h"
+
+/// specializations
+#include "tensorstore/internal/json_binding/absl_time.h"
+#include "tensorstore/internal/json_binding/bindable.h"
+#include "tensorstore/internal/json_binding/json_binding.h"
+#include "tensorstore/internal/json_binding/std_array.h"
+#include "tensorstore/internal/json_binding/std_optional.h"
+
+using ::tensorstore::internal::AnyContextResourceJsonBinder;
+using ::tensorstore::internal::ContextResourceCreationContext;
+using ::tensorstore::internal_kvstore_gcs_http::ScalingRateLimiter;
+using ::tensorstore::internal_kvstore_gcs_http::NoRateLimiter;
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+namespace {
+
+const internal::ContextResourceRegistration<S3RequestRetries>
+    s3_request_retries_registration;
+
+const internal::ContextResourceRegistration<S3ConcurrencyResource>
+    s3_concurrency_registration;
+
+const internal::ContextResourceRegistration<S3RateLimiterResource>
+    s3_rate_limiter_registration;
+
+constexpr size_t kDefaultRequestConcurrency = 32;
+
+std::optional<size_t> GetEnvS3RequestConcurrency() {
+  // Called before flag parsing during resource registration.
+  auto env = internal::GetEnv("TENSORSTORE_S3_REQUEST_CONCURRENCY");
```

Comment:
For these I'm working towards a dual-mode ABSL_FLAG and environment variable to set the values.considering using both flags and env vars in a pattern like I have in gcs_grpc driver:

https://github.com/google/tensorstore/blob/d419261e36f9f33f3b3f8245cdf4bd73a3daf422/tensorstore/kvstore/gcs_grpc/storage_stub_pool.cc#L62

Reply:
Used `ABSL_FLAG` here. I considered using it in `s3_credentials_provider.cc` but with the plethora of options there it feels like the logic would get convoluted fast. Also probably not good practice to pass credentials on the command line.